// @start9labs/start-sdk/package/lib/backup/Backups.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Backups = exports.DEFAULT_OPTIONS = void 0;
const child_process = __importStar(require("child_process"));
const fs = __importStar(require("fs/promises"));
const util_1 = require("../util");
exports.DEFAULT_OPTIONS = {
    delete: true,
    exclude: [],
};
class Backups {
    constructor(options = exports.DEFAULT_OPTIONS, restoreOptions = {}, backupOptions = {}, backupSet = [], preBackup = async (effects) => { }, postBackup = async (effects) => { }, preRestore = async (effects) => { }, postRestore = async (effects) => { }) {
        this.options = options;
        this.restoreOptions = restoreOptions;
        this.backupOptions = backupOptions;
        this.backupSet = backupSet;
        this.preBackup = preBackup;
        this.postBackup = postBackup;
        this.preRestore = preRestore;
        this.postRestore = postRestore;
    }
    static ofVolumes(...volumeNames) {
        return Backups.ofSyncs(...volumeNames.map((srcVolume) => ({
            dataPath: `/media/startos/volumes/${srcVolume}/`,
            backupPath: `/media/startos/backup/volumes/${srcVolume}/`,
        })));
    }
    static ofSyncs(...syncs) {
        return syncs.reduce((acc, x) => acc.addSync(x), new Backups());
    }
    static withOptions(options) {
        return new Backups({ ...exports.DEFAULT_OPTIONS, ...options });
    }
    setOptions(options) {
        this.options = {
            ...this.options,
            ...options,
        };
        return this;
    }
    setBackupOptions(options) {
        this.backupOptions = {
            ...this.backupOptions,
            ...options,
        };
        return this;
    }
    setRestoreOptions(options) {
        this.restoreOptions = {
            ...this.restoreOptions,
            ...options,
        };
        return this;
    }
    setPreBackup(fn) {
        this.preBackup = fn;
        return this;
    }
    setPostBackup(fn) {
        this.postBackup = fn;
        return this;
    }
    setPreRestore(fn) {
        this.preRestore = fn;
        return this;
    }
    setPostRestore(fn) {
        this.postRestore = fn;
        return this;
    }
    addVolume(volume, options) {
        return this.addSync({
            dataPath: `/media/startos/volumes/${volume}/`,
            backupPath: `/media/startos/backup/volumes/${volume}/`,
            ...options,
        });
    }
    addSync(sync) {
        this.backupSet.push(sync);
        return this;
    }
    async createBackup(effects) {
        await this.preBackup(effects);
        for (const item of this.backupSet) {
            const rsyncResults = await runRsync({
                srcPath: item.dataPath,
                dstPath: item.backupPath,
                options: {
                    ...this.options,
                    ...this.backupOptions,
                    ...item.options,
                    ...item.backupOptions,
                },
            });
            await rsyncResults.wait();
        }
        const dataVersion = await effects.getDataVersion();
        if (dataVersion)
            await fs.writeFile("/media/startos/backup/dataVersion.txt", dataVersion, {
                encoding: "utf-8",
            });
        await this.postBackup(effects);
        return;
    }
    async init(effects, kind) {
        if (kind === "restore") {
            await this.restoreBackup(effects);
        }
    }
    async restoreBackup(effects) {
        this.preRestore(effects);
        for (const item of this.backupSet) {
            const rsyncResults = await runRsync({
                srcPath: item.backupPath,
                dstPath: item.dataPath,
                options: {
                    ...this.options,
                    ...this.backupOptions,
                    ...item.options,
                    ...item.backupOptions,
                },
            });
            await rsyncResults.wait();
        }
        const dataVersion = await fs
            .readFile("/media/startos/backup/dataVersion.txt", {
            encoding: "utf-8",
        })
            .catch((_) => null);
        if (dataVersion)
            await effects.setDataVersion({ version: dataVersion });
        this.postRestore(effects);
        return;
    }
}
exports.Backups = Backups;
async function runRsync(rsyncOptions) {
    const { srcPath, dstPath, options } = rsyncOptions;
    await fs.mkdir(dstPath, { recursive: true });
    const command = "rsync";
    const args = [];
    if (options.delete) {
        args.push("--delete");
    }
    for (const exclude of options.exclude) {
        args.push(`--exclude=${exclude}`);
    }
    args.push("-rlptgocAXH");
    args.push("--info=progress2");
    args.push("--no-inc-recursive");
    args.push(srcPath);
    args.push(dstPath);
    const spawned = child_process.spawn(command, args, { detached: true });
    let percentage = 0.0;
    spawned.stdout.on("data", (data) => {
        const lines = String(data).replace(/\r/g, "\n").split("\n");
        for (const line of lines) {
            const parsed = /$([0-9.]+)%/.exec(line)?.[1];
            if (!parsed) {
                console.log(line);
                continue;
            }
            percentage = Number.parseFloat(parsed);
        }
    });
    let stderr = "";
    spawned.stderr.on("data", (data) => {
        const errString = data.toString("utf-8");
        stderr += errString;
        console.error(`Backups.runAsync`, (0, util_1.asError)(errString));
    });
    const id = async () => {
        const pid = spawned.pid;
        if (pid === undefined) {
            throw new Error("rsync process has no pid");
        }
        return String(pid);
    };
    const waitPromise = new Promise((resolve, reject) => {
        spawned.on("exit", (code) => {
            if (code === 0) {
                resolve(null);
            }
            else {
                reject(new Error(`rsync exited with code ${code}\n${stderr}`));
            }
        });
    });
    const wait = () => waitPromise;
    const progress = () => Promise.resolve(percentage);
    return { id, wait, progress };
}
