// ts-matches/lib/parsers/array-of-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayOfParser = void 0;
exports.arrayOf = arrayOf;
const index_1 = require("./index");
/**
 * Given an object, we want to make sure the key exists and that the value on
 * the key matches the parser
 * Note: This will mutate the value sent through
 */
class ArrayOfParser {
    parser;
    description;
    constructor(parser, description = {
        name: "ArrayOf",
        children: [parser],
        extras: [],
    }) {
        this.parser = parser;
        this.description = description;
    }
    parse(a, onParse) {
        if (!Array.isArray(a)) {
            return onParse.invalid({
                value: a,
                keys: [],
                parser: this,
            });
        }
        const values = [...a];
        for (let index = 0; index < values.length; index++) {
            const result = this.parser.enumParsed(values[index]);
            if ("error" in result) {
                result.error.keys.push("" + index);
                return onParse.invalid(result.error);
            }
            else {
                values[index] = result.value;
            }
        }
        return onParse.parsed(values);
    }
}
exports.ArrayOfParser = ArrayOfParser;
/**
 * We would like to validate that all of the array is of the same type
 * @param validator What is the validator for the values in the array
 */
function arrayOf(validator) {
    return new index_1.Parser(new ArrayOfParser(validator));
}
