// ts-matches/lib/parsers/deferred-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeferredParser = void 0;
exports.deferred = deferred;
const parser_1 = require("./parser");
/**
 * This is needed when the typescript has a recursive, mutual types
 * type Things = string | [OtherThings]
 * type OtherThings = {type: 'other', value:Things }
 */
class DeferredParser {
    description;
    parser;
    static create() {
        return new DeferredParser();
    }
    constructor(description = {
        name: "Deferred",
        children: [],
        extras: [],
    }) {
        this.description = description;
    }
    setParser(parser) {
        this.parser = new parser_1.Parser(parser);
        return this;
    }
    parse(a, onParse) {
        if (!this.parser) {
            return onParse.invalid({
                value: "Not Set Up",
                keys: [],
                parser: this,
            });
        }
        return this.parser.parse(a, onParse);
    }
}
exports.DeferredParser = DeferredParser;
/**
 * Must pass the shape that we expect since typescript as of this point
 * can't infer with recursive like structures like this.
 * @returns [Parser, setParser] Use the setParser to set the parser later
 */
function deferred() {
    const deferred = DeferredParser.create();
    function setParser(parser) {
        deferred.setParser(parser);
    }
    return [new parser_1.Parser(deferred), setParser];
}
