// ts-matches/lib/parsers/mapped-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MappedAParser = void 0;
class MappedAParser {
    parent;
    map;
    mappingName;
    description;
    constructor(parent, map, mappingName = map.name, description = {
        name: "Mapped",
        children: [parent],
        extras: [mappingName],
    }) {
        this.parent = parent;
        this.map = map;
        this.mappingName = mappingName;
        this.description = description;
    }
    parse(a, onParse) {
        const map = this.map;
        try {
            const result = this.parent.enumParsed(a);
            if ("error" in result) {
                return onParse.invalid(result.error);
            }
            return onParse.parsed(map(result.value));
        }
        catch (error) {
            return onParse.invalid({
                keys: [String(error)],
                parser: this,
                value: a,
            });
        }
    }
}
exports.MappedAParser = MappedAParser;
