// ts-matches/lib/parsers/simple-parsers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.regex = exports.instanceOf = exports.string = exports.isArray = exports.object = exports.boolean = exports.isFunction = exports.natural = exports.isNill = exports.number = exports.unknown = exports.any = void 0;
exports.guard = guard;
const index_1 = require("./index");
const any_parser_1 = require("./any-parser");
const array_parser_1 = require("./array-parser");
const bool_parser_1 = require("./bool-parser");
const function_parser_1 = require("./function-parser");
const nill_parser_1 = require("./nill-parser");
const number_parser_1 = require("./number-parser");
const object_parser_1 = require("./object-parser");
const string_parser_1 = require("./string-parser");
const unknown_parser_1 = require("./unknown-parser");
/**
 * Create a custom type guard
 * @param test A function that will determine runtime if the value matches
 * @param testName A name for that function, useful when it fails
 */
function guard(test, testName) {
    return index_1.Parser.isA(test, testName || test.name);
}
exports.any = new index_1.Parser(new any_parser_1.AnyParser());
exports.unknown = new index_1.Parser(new unknown_parser_1.UnknownParser());
exports.number = new index_1.Parser(new number_parser_1.NumberParser());
exports.isNill = new index_1.Parser(new nill_parser_1.NilParser());
exports.natural = exports.number.refine((x) => x >= 0 && x === Math.floor(x));
exports.isFunction = new index_1.Parser(new function_parser_1.FunctionParser());
exports.boolean = new index_1.Parser(new bool_parser_1.BoolParser());
const objectMatcher = new index_1.Parser(new object_parser_1.ObjectParser());
exports.object = Object.assign(function objectOf(...args) {
    return index_1.shape(...args);
}, objectMatcher);
exports.isArray = new index_1.Parser(new array_parser_1.ArrayParser());
exports.string = new index_1.Parser(new string_parser_1.StringParser());
const instanceOf = (classCreator) => guard((x) => x instanceof classCreator, `is${classCreator.name}`);
exports.instanceOf = instanceOf;
const regex = (tester) => exports.string.refine(function (x) {
    return tester.test(x);
}, tester.toString());
exports.regex = regex;
