// ts-matches/lib/parsers/tuple-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TupleParser = void 0;
exports.tuple = tuple;
const index_1 = require("./index");
const utils_1 = require("../utils");
class TupleParser {
    parsers;
    lengthMatcher;
    description;
    constructor(parsers, lengthMatcher = (0, index_1.literal)(parsers.length), description = {
        name: "Tuple",
        children: parsers,
        extras: [],
    }) {
        this.parsers = parsers;
        this.lengthMatcher = lengthMatcher;
        this.description = description;
    }
    parse(input, onParse) {
        const tupleError = index_1.isArray.enumParsed(input);
        if ("error" in tupleError)
            return onParse.invalid(tupleError.error);
        const values = input;
        const stateCheck = this.lengthMatcher.enumParsed(values.length);
        if ("error" in stateCheck) {
            stateCheck.error.keys.push((0, utils_1.saferStringify)("length"));
            return onParse.invalid(stateCheck.error);
        }
        const answer = new Array(this.parsers.length);
        for (const key in this.parsers) {
            const parser = this.parsers[key];
            const value = values[key];
            const result = parser.enumParsed(value);
            if ("error" in result) {
                const { error } = result;
                error.keys.push((0, utils_1.saferStringify)(key));
                return onParse.invalid(error);
            }
            answer[key] = result.value;
        }
        return onParse.parsed(answer);
    }
}
exports.TupleParser = TupleParser;
function tuple(...parsers) {
    return new index_1.Parser(new TupleParser(parsers));
}
