// @start9labs/start-sdk/base/lib/actions/input/builder/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.List = void 0;
const ts_matches_1 = require("ts-matches");
class List {
    constructor(build, validator) {
        this.build = build;
        this.validator = validator;
        this._TYPE = null;
    }
    static text(a, aSpec) {
        const validator = (0, ts_matches_1.arrayOf)(ts_matches_1.string);
        return new List(() => {
            const spec = {
                type: "text",
                placeholder: null,
                minLength: null,
                maxLength: null,
                masked: false,
                inputmode: "text",
                generate: null,
                patterns: aSpec.patterns || [],
                ...aSpec,
            };
            const built = {
                description: null,
                warning: null,
                default: [],
                type: "list",
                minLength: null,
                maxLength: null,
                disabled: false,
                ...a,
                spec,
            };
            return { spec: built, validator };
        }, validator);
    }
    static dynamicText(getA) {
        const validator = (0, ts_matches_1.arrayOf)(ts_matches_1.string);
        return new List(async (options) => {
            const { spec: aSpec, ...a } = await getA(options);
            const spec = {
                type: "text",
                placeholder: null,
                minLength: null,
                maxLength: null,
                masked: false,
                inputmode: "text",
                generate: null,
                patterns: aSpec.patterns || [],
                ...aSpec,
            };
            const built = {
                description: null,
                warning: null,
                default: [],
                type: "list",
                minLength: null,
                maxLength: null,
                disabled: false,
                ...a,
                spec,
            };
            return { spec: built, validator };
        }, validator);
    }
    static obj(a, aSpec) {
        return new List(async (options) => {
            const { spec: previousSpecSpec, ...restSpec } = aSpec;
            const built = await previousSpecSpec.build(options);
            const spec = {
                type: "object",
                displayAs: null,
                uniqueBy: null,
                ...restSpec,
                spec: built.spec,
            };
            const value = {
                spec,
                default: [],
                ...a,
            };
            return {
                spec: {
                    description: null,
                    warning: null,
                    minLength: null,
                    maxLength: null,
                    type: "list",
                    disabled: false,
                    ...value,
                },
                validator: (0, ts_matches_1.arrayOf)(built.validator),
            };
        }, (0, ts_matches_1.arrayOf)(aSpec.spec.validator));
    }
}
exports.List = List;
