// @start9labs/start-sdk/base/lib/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UseEntrypoint = exports.ExpectedExports = exports.NO_TIMEOUT = exports.SIGKILL = exports.SIGTERM = exports.inputSpecTypes = void 0;
exports.isUseEntrypoint = isUseEntrypoint;
exports.writable = writable;
exports.readonly = readonly;
exports.inputSpecTypes = __importStar(require("./actions/input/inputSpecTypes"));
__exportStar(require("./osBindings"), exports);
exports.SIGTERM = "SIGTERM";
exports.SIGKILL = "SIGKILL";
exports.NO_TIMEOUT = -1;
var ExpectedExports;
(function (ExpectedExports) {
    version: 1;
})(ExpectedExports || (exports.ExpectedExports = ExpectedExports = {}));
class UseEntrypoint {
    constructor(overridCmd) {
        this.overridCmd = overridCmd;
        this.USE_ENTRYPOINT = "USE_ENTRYPOINT";
    }
}
exports.UseEntrypoint = UseEntrypoint;
function isUseEntrypoint(command) {
    return typeof command === "object" && "USE_ENTRYPOINT" in command;
}
function writable(value) {
    return value;
}
function readonly(value) {
    return value;
}
