// @start9labs/start-sdk/package/lib/version/VersionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionInfo = exports.IMPOSSIBLE = void 0;
exports.IMPOSSIBLE = Symbol("IMPOSSIBLE");
class VersionInfo {
    constructor(options) {
        this.options = options;
        this._version = null;
    }
    /**
     * @description Use this function to define a new version of the service. By convention, each version should receive its own file.
     * @property {string} version
     * @property {string} releaseNotes
     * @property {object} migrations
     * @returns A VersionInfo class instance that is exported, then imported into versions/index.ts.
     */
    static of(options) {
        return new VersionInfo({ ...options, satisfies: [] });
    }
    /** Specify a version that this version is 100% backwards compatible to */
    satisfies(version) {
        return new VersionInfo({
            ...this.options,
            satisfies: [...this.options.satisfies, version],
        });
    }
}
exports.VersionInfo = VersionInfo;
function __type_tests() {
    const version = VersionInfo.of({
        version: "1.0.0:0",
        releaseNotes: "",
        migrations: {},
    })
        .satisfies("#other:1.0.0:0")
        .satisfies("#other:2.0.0:0")
        // @ts-expect-error
        .satisfies("#other:2.f.0:0");
    let a = version;
    // @ts-expect-error
    let b = version;
    VersionInfo.of({
        // @ts-expect-error
        version: "test",
        releaseNotes: "",
        migrations: {},
    });
    VersionInfo.of({
        // @ts-expect-error
        version: "test",
        releaseNotes: "",
        migrations: {},
    });
}
