// @start9labs/start-sdk/package/lib/test/host.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ServiceInterfaceBuilder_1 = require("../../../base/lib/interfaces/ServiceInterfaceBuilder");
const output_sdk_1 = require("../test/output.sdk");
describe("host", () => {
    test("Testing that the types work", () => {
        async function test(effects) {
            const foo = output_sdk_1.sdk.MultiHost.of(effects, "foo");
            const fooOrigin = await foo.bindPort(80, {
                protocol: "http",
                preferredExternalPort: 80,
            });
            const fooInterface = new ServiceInterfaceBuilder_1.ServiceInterfaceBuilder({
                effects,
                name: "Foo",
                id: "foo",
                description: "A Foo",
                type: "ui",
                username: "bar",
                path: "/baz",
                query: { qux: "yes" },
                schemeOverride: null,
                masked: false,
            });
            await fooOrigin.export([fooInterface]);
        }
    });
});
