// ts-matches/lib/matches.deferred.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validatorError = void 0;
const matches_1 = __importDefault(require("./matches"));
const index_1 = require("./parsers/index");
const matches_test_1 = require("./matches.test");
exports.validatorError = (0, index_1.every)((0, index_1.shape)({
    parser: matches_1.default.object,
    keys: matches_1.default.arrayOf(matches_1.default.string),
    value: index_1.any,
}));
const unFold = {
    invalid: index_1.Parser.validatorErrorAsString,
    parsed: (x) => x,
};
{
    const [matchThings, setMatchThings] = (0, index_1.deferred)();
    const matchOtherThings = (0, index_1.shape)({
        type: (0, index_1.literal)("other"),
        value: matchThings,
    });
    setMatchThings((0, index_1.some)(index_1.string, (0, index_1.tuple)(matchOtherThings)));
    test("simple deferred shape", () => {
        const testValue = [
            { type: "other", value: [{ type: "other", value: "test" }] },
        ];
        const parsed = matchThings.unsafeCast(testValue);
        expect(parsed).toEqual(testValue);
        // @ts-expect-error
        (0, matches_test_1.isType)(parsed);
        (0, matches_test_1.isType)(parsed);
    });
    test("simple deferred shape invalid", () => {
        const testValue = [
            { type: "other", value: [{ type: "ot2her", value: "test" }] },
        ];
        try {
            const parsed = matchThings.unsafeCast(testValue);
        }
        catch (e) {
            (0, matches_test_1.assertSnapshot)('"[\\"0\\"][\\"value\\"][\\"0\\"][\\"type\\"]Or<Tuple<Shape<{type:Literal<\\\"other\\\">,value:Deferred<>}>>,string>(\\"ot2her\\")"', matchThings.parse(testValue, unFold));
            return;
        }
        throw new Error("should be invalid");
    });
}
test("deferred failed because we failed to complete ", () => {
    const [matchThings, setMatchThings] = (0, index_1.deferred)();
    const testValue = [
        { type: "other", value: [{ type: "ot2her", value: "test" }] },
    ];
    try {
        const parsed = matchThings.unsafeCast(testValue);
    }
    catch (e) {
        (0, matches_test_1.assertSnapshot)('"Deferred<>(\\"Not Set Up\\")"', matchThings.parse(testValue, unFold));
        return;
    }
    throw new Error("should be invalid");
});
