// ts-matches/lib/matches.gen.helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.testSetup = exports.testSetupInformation = exports.matcherPairs = exports.matcherTuple = exports.matchPairOf = exports.noPossibleCounter = void 0;
const matches_1 = __importDefault(require("./matches"));
const parsers_1 = require("./parsers");
const utils_1 = require("./utils");
const jest_1 = require("@fast-check/jest");
exports.noPossibleCounter = {
    noPossibleCounter: true,
};
const matchPairOf = (matcher, example, type, counterExample) => ({
    matcher,
    type,
    example,
    counterExample,
    toString() {
        return `{
      matcher: ${parsers_1.Parser.parserAsString(this.matcher)},
      example: ${(0, utils_1.saferStringify)(this.example)},
      type: ${this.type},
      counterExample: ${(0, utils_1.saferStringify)(this.counterExample)},
    }`;
    },
});
exports.matchPairOf = matchPairOf;
const matcherPairsSimple = (() => {
    const trueFloat = jest_1.fc
        .tuple(jest_1.fc.oneof(jest_1.fc.constant(0), jest_1.fc.float()), jest_1.fc.integer())
        .map(([x, y]) => x + y);
    const matcherNumber = jest_1.fc
        .record({
        example: trueFloat,
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), jest_1.fc.object()),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.number, example, "number", counterExample));
    const matcherNill = jest_1.fc
        .record({
        example: jest_1.fc.constantFrom(null, undefined),
        counterExample: jest_1.fc.oneof(jest_1.fc.string(), jest_1.fc.object(), jest_1.fc.integer()),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.nill, example, "nill", counterExample));
    const matcherNat = jest_1.fc
        .record({
        example: jest_1.fc.nat(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), jest_1.fc.object(), jest_1.fc
            .nat()
            .filter((x) => x !== 0)
            .map((x) => -x)),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.natural, example, "natural number", counterExample));
    const matcherObject = jest_1.fc
        .record({
        example: jest_1.fc.object(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), trueFloat),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.object, example, "object", counterExample));
    jest_1.fc.boolean(), jest_1.fc.integer(), jest_1.fc.string();
    const matcherFunction = jest_1.fc
        .record({
        example: jest_1.fc.constant(() => { }),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), jest_1.fc.object(), trueFloat),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.isFunction, example, "function", counterExample));
    const matcherBoolean = jest_1.fc
        .record({
        example: jest_1.fc.boolean(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), jest_1.fc.object(), trueFloat),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.boolean, example, "boolean", counterExample));
    const matcherArray = jest_1.fc
        .record({
        example: jest_1.fc.array(jest_1.fc.anything()),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined, {}), jest_1.fc.string(), trueFloat),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.array, example, "array", counterExample));
    const matcherAny = jest_1.fc
        .anything()
        .map((x) => (0, exports.matchPairOf)(matches_1.default.any, x, "any", exports.noPossibleCounter));
    const matcherString = jest_1.fc
        .record({
        example: jest_1.fc.string(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.object(), trueFloat),
    })
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.string, example, "string", counterExample));
    const matcherConstant = jest_1.fc
        .oneof(jest_1.fc.record({
        example: jest_1.fc.boolean(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), trueFloat),
    }), jest_1.fc.record({
        example: jest_1.fc.string(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), trueFloat, jest_1.fc.boolean()),
    }), jest_1.fc.record({
        example: trueFloat,
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), jest_1.fc.boolean()),
    }))
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.literal(example), example, `literal of ${(0, utils_1.saferStringify)(example)}`, counterExample));
    const matchesLiteral = jest_1.fc
        .oneof(jest_1.fc.record({
        example: jest_1.fc.boolean(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), trueFloat),
    }), jest_1.fc.record({
        example: jest_1.fc.string(),
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), trueFloat, jest_1.fc.boolean()),
    }), jest_1.fc.record({
        example: trueFloat,
        counterExample: jest_1.fc.oneof(jest_1.fc.constantFrom(null, undefined), jest_1.fc.string(), jest_1.fc.boolean()),
    }))
        .map(({ example, counterExample }) => (0, exports.matchPairOf)(matches_1.default.literal(example), example, `literal of ${(0, utils_1.saferStringify)(example)}`, counterExample));
    class TestBase {
    }
    const constructors = [
        class Test {
        },
        class Test2 extends TestBase {
        },
        class Test3 {
        },
    ];
    const matcherInstanceOf = jest_1.fc
        .constantFrom(...constructors)
        .chain((Cons) => jest_1.fc.record({
        example: jest_1.fc.constant(new Cons()),
        matcher: jest_1.fc.constant(matches_1.default.instanceOf(Cons)),
        type: jest_1.fc.constant(`instanceOf ${Cons.name}`),
        counterExample: jest_1.fc.constantFrom(...constructors.filter((x) => x != Cons)),
    }))
        .map((base) => (0, exports.matchPairOf)(base.matcher, base.example, base.type, base.counterExample));
    return jest_1.fc.oneof(matcherNumber, matcherFunction, matcherInstanceOf, matcherObject, matcherConstant, matcherBoolean, matchesLiteral, matcherArray, 
    // matcherAny,
    matcherString, matcherNat, matcherNill);
})();
const matcherSome = jest_1.fc
    .array(matcherPairsSimple)
    .filter((x) => x.length > 0)
    .chain((matchers) => jest_1.fc.integer({ min: 0, max: matchers.length - 1 }).map((atIndex) => ({
    atIndex,
    matchers,
})))
    .map(({ atIndex, matchers }) => {
    return (0, exports.matchPairOf)(matches_1.default.some(...matchers.map((x) => x.matcher)), matchers[atIndex].example, `some (${matchers.map((x) => x.type).join(", ")})`, exports.noPossibleCounter);
});
const matcherArrayOf = matcherPairsSimple.chain((pair) => jest_1.fc.array(jest_1.fc.constant(pair)).map((pairs) => {
    const validCounter = !!pairs.length &&
        pairs.every((x) => x.counterExample !== exports.noPossibleCounter);
    return (0, exports.matchPairOf)(matches_1.default.arrayOf(pair.matcher), pairs.map((x) => x.example), `arrayOf ${pair.type}`, validCounter ? pairs.map((x) => x.counterExample) : null);
}));
exports.matcherTuple = jest_1.fc.array(matcherPairsSimple).map((xs) => {
    const validCounter = !!xs.length && xs.every((x) => x.counterExample !== exports.noPossibleCounter);
    return (0, exports.matchPairOf)(matches_1.default.tuple(...xs.map((tupleValue) => tupleValue.matcher)), xs.map((x) => x.example), `tuple ${(0, utils_1.saferStringify)(xs.map((x) => x.type))}`, validCounter ? xs.map((x) => x.counterExample) : 0);
});
const matcherShape = jest_1.fc.dictionary(jest_1.fc.string(), matcherPairsSimple).map((x) => {
    const matcher = matches_1.default.shape(Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.matcher;
        return acc;
    }, {}));
    const example = Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.example;
        return acc;
    }, {});
    const type = `shape of ${(0, utils_1.saferStringify)(Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.type;
        return acc;
    }, {}))}`;
    const counterExample = Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.counterExample;
        return acc;
    }, {});
    const validCounter = !!Object.keys(x).length &&
        Object.values(x).every((x) => x.counterExample !== exports.noPossibleCounter);
    return (0, exports.matchPairOf)(matcher, example, type, validCounter ? counterExample : 0);
});
const matcherShapePartial = jest_1.fc
    .dictionary(jest_1.fc.string(), matcherPairsSimple)
    .map((x) => {
    const matcher = matches_1.default.partial(Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.matcher;
        return acc;
    }, {}));
    const example = Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.example;
        return acc;
    }, {});
    const type = `shape of ${(0, utils_1.saferStringify)(Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.type;
        return acc;
    }, {}))}`;
    const validCounter = !!Object.keys(x).length &&
        Object.values(x).every((x) => x.counterExample !== exports.noPossibleCounter);
    const counterExample = Object.keys(x).reduce((acc, key) => {
        const value = x[key];
        acc[key] = value.counterExample;
        return acc;
    }, {});
    return (0, exports.matchPairOf)(matcher, example, type, validCounter ? counterExample : 0);
});
exports.matcherPairs = jest_1.fc.oneof(matcherPairsSimple, matcherShape, exports.matcherTuple, matcherArrayOf, matcherShapePartial, matcherSome);
const testSetupInformation = (matchPair) => ({
    ...matchPair,
    matchValue: {},
});
exports.testSetupInformation = testSetupInformation;
exports.testSetup = jest_1.fc.array(exports.matcherPairs).chain((matcherPairsSets) => {
    const defaultValue = {};
    const defaultTest = {
        defaultValue,
        setupInformation: [],
        runMatch: (x) => (0, matches_1.default)(x).defaultTo(-1),
        runMatchLazy: (x) => (0, matches_1.default)(x).defaultToLazy(() => -1),
        randomExample: {
            value: defaultValue,
            counter: exports.noPossibleCounter,
            index: -1,
        },
        toString() {
            return `{
        matching: ${(0, utils_1.saferStringify)(matcherPairsSets.map((x) => x.matcher).map((x) => parsers_1.Parser.parserAsString(x)))},
        defaultValue: ${(0, utils_1.saferStringify)(this.defaultValue)},
        randomExample: {
          value: ${(0, utils_1.saferStringify)(this.randomExample.value)},
          counter: ${(0, utils_1.saferStringify)(this.randomExample.counter)},
          index: ${this.randomExample.index},
        },
      }`;
        },
    };
    if (!matcherPairsSets.length) {
        return jest_1.fc.constant(defaultTest);
    }
    const setupInformation = matcherPairsSets.reduce((acc, value) => {
        acc.push((0, exports.testSetupInformation)(value));
        return acc;
    }, []);
    return jest_1.fc.nat(matcherPairsSets.length).map((randomExampleIndex) => {
        if (randomExampleIndex === matcherPairsSets.length) {
            return defaultTest;
        }
        const randomExample = {
            value: setupInformation[randomExampleIndex].example,
            counter: setupInformation[randomExampleIndex].counterExample,
            index: randomExampleIndex,
        };
        return {
            defaultValue,
            setupInformation,
            runMatch: (x) => setupInformation
                .reduce((acc, { matcher, matchValue }, index) => acc.when(matcher, () => index), (0, matches_1.default)(x))
                .defaultTo(-1),
            runMatchLazy: (x) => setupInformation
                .reduce((acc, { matcher, matchValue }, index) => acc.when(matcher, () => index), (0, matches_1.default)(x))
                .defaultToLazy(() => -1),
            randomExample,
            toString() {
                return `{
          matching: ${(0, utils_1.saferStringify)(matcherPairsSets.map((x) => parsers_1.Parser.parserAsString(x.matcher)))},
          defaultValue: ${(0, utils_1.saferStringify)(defaultValue)},
          randomExample: {
            value: ${(0, utils_1.saferStringify)(randomExample.value)},
            counter: ${(0, utils_1.saferStringify)(randomExample.counter)},
            index: ${randomExample.index},
          },
        }`;
            },
        };
    });
});
