// @start9labs/start-sdk/base/lib/actions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTask = exports.runAction = void 0;
const runAction = async (options) => {
    if (options.input) {
        if (options.input instanceof Function) {
            const prev = await options.effects.action.getInput({
                // packageId: options.packageId,
                actionId: options.actionId,
            });
            const input = options.input(prev
                ? { spec: prev.spec, value: prev.value }
                : undefined);
            return options.effects.action.run({
                // packageId: options.packageId,
                actionId: options.actionId,
                input,
            });
        }
        else {
            return options.effects.action.run({
                // packageId: options.packageId,
                actionId: options.actionId,
                input: options.input,
            });
        }
    }
    else {
        return options.effects.action.run({
            //   packageId: options.packageId,
            actionId: options.actionId,
        });
    }
};
exports.runAction = runAction;
const _validate = {};
const createTask = (options) => {
    const request = options.options || {};
    const actionId = options.action.id;
    const req = {
        ...request,
        actionId,
        packageId: options.packageId,
        action: undefined,
        severity: options.severity,
        replayId: request.replayId || `${options.packageId}:${actionId}`,
    };
    delete req.action;
    return options.effects.action.createTask(req);
};
exports.createTask = createTask;
