// @start9labs/start-sdk/base/lib/actions/input/inputSpecConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.smtpInputSpec = exports.customSmtp = void 0;
const util_1 = require("../../util");
const inputSpec_1 = require("./builder/inputSpec");
const value_1 = require("./builder/value");
const variants_1 = require("./builder/variants");
/**
 * Base SMTP settings, to be used by StartOS for system wide SMTP
 */
exports.customSmtp = inputSpec_1.InputSpec.of({
    server: value_1.Value.text({
        name: "SMTP Server",
        required: true,
        default: null,
    }),
    port: value_1.Value.number({
        name: "Port",
        required: true,
        default: 587,
        min: 1,
        max: 65535,
        integer: true,
    }),
    from: value_1.Value.text({
        name: "From Address",
        required: true,
        default: null,
        placeholder: "Example Name <test@example.com>",
        inputmode: "email",
        patterns: [util_1.Patterns.emailWithName],
    }),
    login: value_1.Value.text({
        name: "Login",
        required: true,
        default: null,
    }),
    password: value_1.Value.text({
        name: "Password",
        required: false,
        default: null,
        masked: true,
    }),
});
const smtpVariants = variants_1.Variants.of({
    disabled: { name: "Disabled", spec: inputSpec_1.InputSpec.of({}) },
    system: {
        name: "System Credentials",
        spec: inputSpec_1.InputSpec.of({
            customFrom: value_1.Value.text({
                name: "Custom From Address",
                description: "A custom from address for this service. If not provided, the system from address will be used.",
                required: false,
                default: null,
                placeholder: "<name>test@example.com",
                inputmode: "email",
                patterns: [util_1.Patterns.email],
            }),
        }),
    },
    custom: {
        name: "Custom Credentials",
        spec: exports.customSmtp,
    },
});
/**
 * For service inputSpec. Gives users 3 options for SMTP: (1) disabled, (2) use system SMTP settings, (3) use custom SMTP settings
 */
exports.smtpInputSpec = value_1.Value.dynamicUnion(async ({ effects }) => {
    const smtp = await new util_1.GetSystemSmtp(effects).once();
    const disabled = smtp ? [] : ["system"];
    return {
        name: "SMTP",
        description: "Optionally provide an SMTP server for sending emails",
        default: "disabled",
        disabled,
        variants: smtpVariants,
    };
}, smtpVariants.validator);
