// @start9labs/start-sdk/package/lib/mainFn/HealthDaemon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HealthDaemon = exports.EXIT_SUCCESS = void 0;
const defaultTrigger_1 = require("../trigger/defaultTrigger");
const oncePromise = () => {
    let resolve;
    const promise = new Promise((res) => {
        resolve = res;
    });
    return { resolve: resolve, promise };
};
exports.EXIT_SUCCESS = "EXIT_SUCCESS";
/**
 * Wanted a structure that deals with controlling daemons by their health status
 * States:
 * -- Waiting for dependencies to be success
 * -- Running: Daemon is running and the status is in the health
 *
 */
class HealthDaemon {
    constructor(daemon, dependencies, id, ready, effects) {
        this.daemon = daemon;
        this.dependencies = dependencies;
        this.id = id;
        this.ready = ready;
        this.effects = effects;
        this._health = { result: "waiting", message: null };
        this.healthWatchers = [];
        this.running = false;
        this.resolvedReady = false;
        this.healthCheckCleanup = null;
        this.readyPromise = new Promise((resolve) => (this.resolveReady = () => {
            resolve();
            this.resolvedReady = true;
        }));
        this.dependencies.forEach((d) => d.addWatcher(() => this.updateStatus()));
    }
    /** Run after we want to do cleanup */
    async term(termOptions) {
        this.healthWatchers = [];
        this.running = false;
        this.healthCheckCleanup?.();
        await this.daemon?.term({
            ...termOptions,
        });
    }
    /** Want to add another notifier that the health might have changed */
    addWatcher(watcher) {
        this.healthWatchers.push(watcher);
    }
    get health() {
        return Object.freeze(this._health);
    }
    async changeRunning(newStatus) {
        if (this.running === newStatus)
            return;
        this.running = newStatus;
        if (newStatus) {
            console.debug(`Launching ${this.id}...`);
            this.setupHealthCheck();
            this.daemon?.start();
            this.started = performance.now();
        }
        else {
            console.debug(`Stopping ${this.id}...`);
            this.daemon?.term();
            await this.turnOffHealthCheck();
        }
    }
    async turnOffHealthCheck() {
        await this.healthCheckCleanup?.();
        this.resolvedReady = false;
        this.readyPromise = new Promise((resolve) => (this.resolveReady = () => {
            resolve();
            this.resolvedReady = true;
        }));
    }
    async setupHealthCheck() {
        this.daemon?.onExit((success) => {
            if (success && this.ready === "EXIT_SUCCESS") {
                this.setHealth({ result: "success", message: null });
            }
            else if (!success) {
                this.setHealth({
                    result: "failure",
                    message: `${this.id} daemon crashed`,
                });
            }
            else if (!this.daemon?.isOneshot()) {
                this.setHealth({
                    result: "failure",
                    message: `${this.id} daemon exited`,
                });
            }
        });
        if (this.ready === "EXIT_SUCCESS")
            return;
        if (this.healthCheckCleanup)
            return;
        const trigger = (this.ready.trigger ?? defaultTrigger_1.defaultTrigger)(() => ({
            lastResult: this._health.result,
        }));
        const { promise: status, resolve: setStatus } = oncePromise();
        const { promise: exited, resolve: setExited } = oncePromise();
        new Promise(async () => {
            if (this.ready === "EXIT_SUCCESS")
                return;
            for (let res = await Promise.race([status, trigger.next()]); !res.done; res = await Promise.race([status, trigger.next()])) {
                const response = await Promise.resolve(this.ready.fn()).catch((err) => {
                    return {
                        result: "failure",
                        message: "message" in err ? err.message : String(err),
                    };
                });
                await this.setHealth(response);
            }
            setExited(null);
        }).catch((err) => console.error(`Daemon ${this.id} failed: ${err}`));
        this.healthCheckCleanup = async () => {
            setStatus({ done: true });
            await exited;
            this.healthCheckCleanup = null;
            return null;
        };
    }
    onReady() {
        return this.readyPromise;
    }
    get isReady() {
        return this.resolvedReady;
    }
    async setHealth(health) {
        const changed = this._health.result !== health.result;
        this._health = health;
        if (this.resolveReady && health.result === "success") {
            this.resolveReady();
        }
        if (changed)
            this.healthWatchers.forEach((watcher) => watcher());
        if (this.ready === "EXIT_SUCCESS")
            return;
        const display = this.ready.display;
        if (!display) {
            return;
        }
        let result = health.result;
        if (result === "failure" &&
            this.started &&
            performance.now() - this.started <= (this.ready.gracePeriod ?? 10_000))
            result = "starting";
        if (result === "failure") {
            console.error(`Health Check ${this.id} failed:`, health.message);
        }
        await this.effects.setHealth({
            ...health,
            id: this.id,
            name: display,
            result,
        });
    }
    async updateStatus() {
        const healths = this.dependencies.map((d) => ({
            health: d.running && d._health,
            id: d.id,
            display: typeof d.ready === "object" ? d.ready.display : null,
        }));
        const waitingOn = healths.filter((h) => !h.health || h.health.result !== "success");
        if (waitingOn.length)
            console.debug(`daemon ${this.id} waiting on ${waitingOn.map((w) => w.id)}`);
        if (waitingOn.length) {
            const waitingOnNames = waitingOn.flatMap((w) => w.display ? [w.display] : []);
            const message = waitingOnNames.length ? waitingOnNames.join(", ") : null;
            await this.setHealth({ result: "waiting", message });
        }
        else {
            await this.setHealth({ result: "starting", message: null });
        }
        await this.changeRunning(!waitingOn.length);
    }
}
exports.HealthDaemon = HealthDaemon;
