// @start9labs/start-sdk/base/lib/util/getServiceInterfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetServiceInterfaces = void 0;
exports.getOwnServiceInterfaces = getOwnServiceInterfaces;
exports.getServiceInterfaces = getServiceInterfaces;
const deepEqual_1 = require("./deepEqual");
const Drop_1 = require("./Drop");
const getServiceInterface_1 = require("./getServiceInterface");
const makeManyInterfaceFilled = async ({ effects, packageId, callback, }) => {
    const serviceInterfaceValues = await effects.listServiceInterfaces({
        packageId,
        callback,
    });
    const serviceInterfacesFilled = await Promise.all(Object.values(serviceInterfaceValues).map(async (serviceInterfaceValue) => {
        const hostId = serviceInterfaceValue.addressInfo.hostId;
        const host = await effects.getHostInfo({
            packageId,
            hostId,
            callback,
        });
        if (!host) {
            throw new Error(`host ${hostId} not found!`);
        }
        return {
            ...serviceInterfaceValue,
            host,
            addressInfo: (0, getServiceInterface_1.filledAddress)(host, serviceInterfaceValue.addressInfo),
        };
    }));
    return serviceInterfacesFilled;
};
class GetServiceInterfaces {
    constructor(effects, opts, map, eq) {
        this.effects = effects;
        this.opts = opts;
        this.map = map;
        this.eq = eq;
    }
    /**
     * Returns the service interfaces for the package. Reruns the context from which it has been called if the underlying value changes
     */
    async const() {
        let abort = new AbortController();
        const watch = this.watch(abort.signal);
        const res = await watch.next();
        if (this.effects.constRetry) {
            watch
                .next()
                .then(() => {
                abort.abort();
                this.effects.constRetry && this.effects.constRetry();
            })
                .catch();
        }
        return res.value;
    }
    /**
     * Returns the service interfaces for the package. Does nothing if the value changes
     */
    async once() {
        const { packageId } = this.opts;
        const interfaceFilled = await makeManyInterfaceFilled({
            effects: this.effects,
            packageId,
        });
        return this.map(interfaceFilled);
    }
    async *watchGen(abort) {
        let prev = null;
        const { packageId } = this.opts;
        const resolveCell = { resolve: () => { } };
        this.effects.onLeaveContext(() => {
            resolveCell.resolve();
        });
        abort?.addEventListener("abort", () => resolveCell.resolve());
        while (this.effects.isInContext && !abort?.aborted) {
            let callback = () => { };
            const waitForNext = new Promise((resolve) => {
                callback = resolve;
                resolveCell.resolve = resolve;
            });
            const next = this.map(await makeManyInterfaceFilled({
                effects: this.effects,
                packageId,
                callback,
            }));
            if (!prev || !this.eq(prev.value, next)) {
                yield next;
            }
            await waitForNext;
        }
        return new Promise((_, rej) => rej(new Error("aborted")));
    }
    /**
     * Watches the service interfaces for the package. Returns an async iterator that yields whenever the value changes
     */
    watch(abort) {
        const ctrl = new AbortController();
        abort?.addEventListener("abort", () => ctrl.abort());
        return Drop_1.DropGenerator.of(this.watchGen(ctrl.signal), () => ctrl.abort());
    }
    /**
     * Watches the service interfaces for the package. Takes a custom callback function to run whenever the value changes
     */
    onChange(callback) {
        ;
        (async () => {
            const ctrl = new AbortController();
            for await (const value of this.watch(ctrl.signal)) {
                try {
                    const res = await callback(value);
                    if (res.cancel) {
                        ctrl.abort();
                        break;
                    }
                }
                catch (e) {
                    console.error("callback function threw an error @ GetServiceInterfaces.onChange", e);
                }
            }
        })()
            .catch((e) => callback(null, e))
            .catch((e) => console.error("callback function threw an error @ GetServiceInterfaces.onChange", e));
    }
    /**
     * Watches the service interfaces for the package. Returns when the predicate is true
     */
    waitFor(pred) {
        const ctrl = new AbortController();
        return Drop_1.DropPromise.of(Promise.resolve().then(async () => {
            for await (const next of this.watchGen(ctrl.signal)) {
                if (pred(next)) {
                    return next;
                }
            }
            throw new Error("context left before predicate passed");
        }), () => ctrl.abort());
    }
}
exports.GetServiceInterfaces = GetServiceInterfaces;
function getOwnServiceInterfaces(effects, map, eq) {
    return new GetServiceInterfaces(effects, {}, map ?? ((a) => a), eq ?? ((a, b) => (0, deepEqual_1.deepEqual)(a, b)));
}
function getServiceInterfaces(effects, opts, map, eq) {
    return new GetServiceInterfaces(effects, opts, map ?? ((a) => a), eq ?? ((a, b) => (0, deepEqual_1.deepEqual)(a, b)));
}
