// @start9labs/start-sdk/base/lib/util/getServiceInterface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetServiceInterface = exports.filledAddress = exports.addressHostToUrl = exports.getHostname = void 0;
exports.getOwnServiceInterface = getOwnServiceInterface;
exports.getServiceInterface = getServiceInterface;
const Host_1 = require("../interfaces/Host");
const Drop_1 = require("./Drop");
const ip_1 = require("./ip");
const deepEqual_1 = require("./deepEqual");
const once_1 = require("./once");
const getHostnameRegex = /^(\w+:\/\/)?([^\/\:]+)(:\d{1,3})?(\/)?/;
const getHostname = (url) => {
    const founds = url.match(getHostnameRegex)?.[2];
    if (!founds)
        return null;
    const parts = founds.split("@");
    const last = parts[parts.length - 1];
    return last;
};
exports.getHostname = getHostname;
const nonLocalFilter = {
    exclude: {
        kind: ["localhost", "link-local"],
    },
};
const publicFilter = {
    visibility: "public",
};
const onionFilter = {
    kind: "onion",
};
const either = (...args) => (a) => args.some((x) => x(a));
const negate = (fn) => (a) => !fn(a);
const unique = (values) => Array.from(new Set(values));
const addressHostToUrl = ({ scheme, sslScheme, username, suffix }, hostname) => {
    const res = [];
    const fmt = (scheme, host, port) => {
        const excludePort = scheme &&
            scheme in Host_1.knownProtocols &&
            port === Host_1.knownProtocols[scheme].defaultPort;
        let hostname;
        if (host.kind === "onion") {
            hostname = host.hostname.value;
        }
        else if (host.kind === "ip") {
            if (host.hostname.kind === "domain") {
                hostname = host.hostname.value;
            }
            else if (host.hostname.kind === "ipv6") {
                hostname = ip_1.IPV6_LINK_LOCAL.contains(host.hostname.value)
                    ? `[${host.hostname.value}%${host.hostname.scopeId}]`
                    : `[${host.hostname.value}]`;
            }
            else {
                hostname = host.hostname.value;
            }
        }
        return `${scheme ? `${scheme}://` : ""}${username ? `${username}@` : ""}${hostname}${excludePort ? "" : `:${port}`}${suffix}`;
    };
    let url = null;
    if (hostname.hostname.port !== null) {
        url = fmt(scheme, hostname, hostname.hostname.port);
    }
    let sslUrl = null;
    if (hostname.hostname.sslPort !== null) {
        sslUrl = fmt(sslScheme, hostname, hostname.hostname.sslPort);
    }
    return { url, sslUrl };
};
exports.addressHostToUrl = addressHostToUrl;
function filterRec(hostnames, filter, invert) {
    if (filter.predicate) {
        const pred = filter.predicate;
        hostnames = hostnames.filter((h) => invert !== pred(h));
    }
    if (filter.visibility === "public")
        hostnames = hostnames.filter((h) => invert !== (h.kind === "onion" || h.public));
    if (filter.visibility === "private")
        hostnames = hostnames.filter((h) => invert !== (h.kind !== "onion" && !h.public));
    if (filter.kind) {
        const kind = new Set(Array.isArray(filter.kind) ? filter.kind : [filter.kind]);
        if (kind.has("ip")) {
            kind.add("ipv4");
            kind.add("ipv6");
        }
        hostnames = hostnames.filter((h) => invert !==
            ((kind.has("onion") && h.kind === "onion") ||
                (kind.has("mdns") &&
                    h.kind === "ip" &&
                    h.hostname.kind === "local") ||
                (kind.has("domain") &&
                    h.kind === "ip" &&
                    h.hostname.kind === "domain") ||
                (kind.has("ipv4") && h.kind === "ip" && h.hostname.kind === "ipv4") ||
                (kind.has("ipv6") && h.kind === "ip" && h.hostname.kind === "ipv6") ||
                (kind.has("localhost") &&
                    ["localhost", "127.0.0.1", "::1"].includes(h.hostname.value)) ||
                (kind.has("link-local") &&
                    h.kind === "ip" &&
                    h.hostname.kind === "ipv6" &&
                    ip_1.IPV6_LINK_LOCAL.contains(ip_1.IpAddress.parse(h.hostname.value)))));
    }
    if (filter.exclude)
        return filterRec(hostnames, filter.exclude, !invert);
    return hostnames;
}
const filledAddress = (host, addressInfo) => {
    const toUrls = exports.addressHostToUrl.bind(null, addressInfo);
    const toUrlArray = (h) => {
        const u = toUrls(h);
        return [u.url, u.sslUrl].filter((u) => u !== null);
    };
    const hostnames = host.hostnameInfo[addressInfo.internalPort] ?? [];
    function filledAddressFromHostnames(hostnames) {
        const getNonLocal = (0, once_1.once)(() => filledAddressFromHostnames(filterRec(hostnames, nonLocalFilter, false)));
        const getPublic = (0, once_1.once)(() => filledAddressFromHostnames(filterRec(hostnames, publicFilter, false)));
        const getOnion = (0, once_1.once)(() => filledAddressFromHostnames(filterRec(hostnames, onionFilter, false)));
        return {
            ...addressInfo,
            hostnames,
            toUrls,
            format: (format) => {
                let res = hostnames;
                if (format === "hostname-info")
                    return res;
                const urls = hostnames.flatMap(toUrlArray);
                if (format === "url")
                    res = urls.map((u) => new URL(u));
                else
                    res = urls;
                return res;
            },
            filter: (filter) => {
                return filledAddressFromHostnames(filterRec(hostnames, filter, false));
            },
            get nonLocal() {
                return getNonLocal();
            },
            get public() {
                return getPublic();
            },
            get onion() {
                return getOnion();
            },
        };
    }
    return filledAddressFromHostnames(hostnames);
};
exports.filledAddress = filledAddress;
const makeInterfaceFilled = async ({ effects, id, packageId, callback, }) => {
    const serviceInterfaceValue = await effects.getServiceInterface({
        serviceInterfaceId: id,
        packageId,
        callback,
    });
    if (!serviceInterfaceValue) {
        return null;
    }
    const hostId = serviceInterfaceValue.addressInfo.hostId;
    const host = await effects.getHostInfo({
        packageId,
        hostId,
        callback,
    });
    const interfaceFilled = {
        ...serviceInterfaceValue,
        host,
        addressInfo: host
            ? (0, exports.filledAddress)(host, serviceInterfaceValue.addressInfo)
            : null,
    };
    return interfaceFilled;
};
class GetServiceInterface {
    constructor(effects, opts, map, eq) {
        this.effects = effects;
        this.opts = opts;
        this.map = map;
        this.eq = eq;
    }
    /**
     * Returns the requested service interface. Reruns the context from which it has been called if the underlying value changes
     */
    async const() {
        let abort = new AbortController();
        const watch = this.watch(abort.signal);
        const res = await watch.next();
        if (this.effects.constRetry) {
            watch
                .next()
                .then(() => {
                abort.abort();
                this.effects.constRetry && this.effects.constRetry();
            })
                .catch();
        }
        return res.value;
    }
    /**
     * Returns the requested service interface. Does nothing if the value changes
     */
    async once() {
        const { id, packageId } = this.opts;
        const interfaceFilled = await makeInterfaceFilled({
            effects: this.effects,
            id,
            packageId,
        });
        return this.map(interfaceFilled);
    }
    async *watchGen(abort) {
        let prev = null;
        const { id, packageId } = this.opts;
        const resolveCell = { resolve: () => { } };
        this.effects.onLeaveContext(() => {
            resolveCell.resolve();
        });
        abort?.addEventListener("abort", () => resolveCell.resolve());
        while (this.effects.isInContext && !abort?.aborted) {
            let callback = () => { };
            const waitForNext = new Promise((resolve) => {
                callback = resolve;
                resolveCell.resolve = resolve;
            });
            const next = this.map(await makeInterfaceFilled({
                effects: this.effects,
                id,
                packageId,
                callback,
            }));
            if (!prev || !this.eq(prev.value, next)) {
                yield next;
            }
            await waitForNext;
        }
        return new Promise((_, rej) => rej(new Error("aborted")));
    }
    /**
     * Watches the requested service interface. Returns an async iterator that yields whenever the value changes
     */
    watch(abort) {
        const ctrl = new AbortController();
        abort?.addEventListener("abort", () => ctrl.abort());
        return Drop_1.DropGenerator.of(this.watchGen(ctrl.signal), () => ctrl.abort());
    }
    /**
     * Watches the requested service interface. Takes a custom callback function to run whenever the value changes
     */
    onChange(callback) {
        ;
        (async () => {
            const ctrl = new AbortController();
            for await (const value of this.watch(ctrl.signal)) {
                try {
                    const res = await callback(value);
                    if (res.cancel) {
                        ctrl.abort();
                        break;
                    }
                }
                catch (e) {
                    console.error("callback function threw an error @ GetServiceInterface.onChange", e);
                }
            }
        })()
            .catch((e) => callback(null, e))
            .catch((e) => console.error("callback function threw an error @ GetServiceInterface.onChange", e));
    }
    /**
     * Watches the requested service interface. Returns when the predicate is true
     */
    waitFor(pred) {
        const ctrl = new AbortController();
        return Drop_1.DropPromise.of(Promise.resolve().then(async () => {
            for await (const next of this.watchGen(ctrl.signal)) {
                if (pred(next)) {
                    return next;
                }
            }
            throw new Error("context left before predicate passed");
        }), () => ctrl.abort());
    }
}
exports.GetServiceInterface = GetServiceInterface;
function getOwnServiceInterface(effects, id, map, eq) {
    return new GetServiceInterface(effects, { id }, map ?? ((a) => a), eq ?? ((a, b) => (0, deepEqual_1.deepEqual)(a, b)));
}
function getServiceInterface(effects, opts, map, eq) {
    return new GetServiceInterface(effects, opts, map ?? ((a) => a), eq ?? ((a, b) => (0, deepEqual_1.deepEqual)(a, b)));
}
